File: deepeval/models/providers/ollama_model.py
from ollama import Client, AsyncClient, ChatResponse
from typing import Optional, Tuple, Union, Dict
from pydantic import BaseModel

from deepeval.models import DeepEvalBaseLLM
from deepeval.key_handler import KeyValues, KEY_FILE_HANDLER


class OllamaModel(DeepEvalBaseLLM):
    def __init__(
        self,
    ):
        model_name = KEY_FILE_HANDLER.fetch_data(KeyValues.LOCAL_MODEL_NAME)
        self.base_url = KEY_FILE_HANDLER.fetch_data(
            KeyValues.LOCAL_EMBEDDING_BASE_URL
        )
        super().__init__(model_name)

    ###############################################
    # Other generate functions
    ###############################################

    def generate(
        self, prompt: str, schema: Optional[BaseModel] = None
    ) -> Tuple[Union[str, Dict], float]:
        chat_model = self.load_model()
        response: ChatResponse = chat_model.chat(
            model=self.model_name,
            messages=[{"role": "user", "content": prompt}],
            format=schema.model_json_schema() if schema else None,
        )
        return (
            (
                schema.model_validate_json(response.message.content)
                if schema
                else response.message.content
            ),
            0,
        )

    async def a_generate(
        self, prompt: str, schema: Optional[BaseModel] = None
    ) -> Tuple[str, float]:
        chat_model = self.load_model(async_mode=True)
        model_name = KEY_FILE_HANDLER.fetch_data(KeyValues.LOCAL_MODEL_NAME)
        response: ChatResponse = await chat_model.chat(
            model=model_name,
            messages=[{"role": "user", "content": prompt}],
            format=schema.model_json_schema() if schema else None,
        )
        return (
            (
                schema.model_validate_json(response.message.content)
                if schema
                else response.message.content
            ),
            0,
        )

    ###############################################
    # Model
    ###############################################

    def load_model(self, async_mode: bool = False):
        if not async_mode:
            return Client(host=self.base_url)
        else:
            return AsyncClient(host=self.base_url)

    def get_model_name(self):
        return f"{self.model_name} (Ollama)"

File: typescript/src/models/providers/ollama-model.ts
import type { ZodType } from "zod";
import {
  DeepEvalBaseLLM,
  type ExtraGenerationParams,
  type GenerationResult,
} from "@/models/base-model";
import { extractJson, importOptional, toJsonSchema } from "@/models/utils";
import { ollamaMessages } from "@/models/multimodal";
import type { ModelNamespace } from "@/models/registry";

const DEFAULT_OLLAMA_BASE_URL = "http://localhost:11434";

/** Any other key is merged into Ollama's `chat(..., options: {...})` bag. */
export interface OllamaModelOptions extends ExtraGenerationParams {
  model?: string;
  baseURL?: string;
  /** Defaults to `0`. Pass `null` to omit it from the request entirely. */
  temperature?: number | null;
  costPerInputToken?: number;
  costPerOutputToken?: number;
}

export class OllamaModel extends DeepEvalBaseLLM {
  private readonly baseURL: string;
  private readonly extraParams: ExtraGenerationParams;
  private client?: any;
  protected registryNamespace: ModelNamespace = "ollama";

  constructor(options: OllamaModelOptions = {}) {
    const {
      model,
      baseURL,
      temperature,
      costPerInputToken,
      costPerOutputToken,
      ...extraParams
    } = options;

    super(model ?? process.env.OLLAMA_MODEL_NAME);
    this.baseURL =
      baseURL ?? process.env.LOCAL_MODEL_BASE_URL ?? DEFAULT_OLLAMA_BASE_URL;
    this.temperature = temperature;
    this.costPerInputToken = costPerInputToken;
    this.costPerOutputToken = costPerOutputToken;
    this.extraParams = extraParams;
  }

  private async getClient(): Promise<any> {
    if (!this.client) {
      const { Ollama } = await importOptional("ollama", "Ollama");
      this.client = new Ollama({ host: this.baseURL });
    }
    return this.client;
  }

  async generate<T = string>(
    prompt: string,
    schema?: ZodType<T>,
  ): Promise<GenerationResult<T>> {
    if (!this.modelName) {
      throw new Error(
        "Ollama requires a model name. Pass `model` or set OLLAMA_MODEL_NAME.",
      );
    }
    const client = await this.getClient();

    const temperature = this.resolveTemperature();
    const modelOptions: Record<string, unknown> = {
      ...(temperature !== undefined && { temperature }),
      ...this.extraParams,
    };

    const request: Record<string, unknown> = {
      model: this.modelName,
      messages: await ollamaMessages(prompt),
      ...(Object.keys(modelOptions).length > 0 && { options: modelOptions }),
    };
    if (schema) {
      request.format = toJsonSchema(schema);
    }

    const response = await client.chat(request);
    const text: string = response.message?.content ?? "";
    const cost = this.resolveCost(
      response.prompt_eval_count,
      response.eval_count,
    );

    if (schema) {
      return { output: schema.parse(extractJson(text)), cost };
    }
    return { output: text as T, cost };
  }

  getModelName(): string {
    return this.modelName ?? "ollama";
  }
}
